rand_distr/src/gamma.rs

```diff
@@ -542,31 +572,144 @@ impl std::error::Error for BetaError {}
 impl<F> Beta<F>
 where
     F: Float,
-    StandardNormal: Distribution<F>,
-    Exp1: Distribution<F>,
     Open01: Distribution<F>,
 {
     /// Construct an object representing the `Beta(alpha, beta)`
     /// distribution.
     pub fn new(alpha: F, beta: F) -> Result<Beta<F>, BetaError> {
-        Ok(Beta {
-            gamma_a: Gamma::new(alpha, F::one()).map_err(|_| BetaError::AlphaTooSmall)?,
-            gamma_b: Gamma::new(beta, F::one()).map_err(|_| BetaError::BetaTooSmall)?,
-        })
+        if !(alpha > F::zero()) {
+            return Err(BetaError::AlphaTooSmall);
+        }
+        if !(beta > F::zero()) {
+            return Err(BetaError::BetaTooSmall);
+        }
+        // From now on, we use the notation from the reference,
+        // i.e. `alpha` and `beta` are renamed to `a0` and `b0`.
+        let (a0, b0) = (alpha, beta);
+        let (a, b, switched_params) = if a0 < b0 {
+            (a0, b0, false)
+        } else {
+            (b0, a0, true)
+        };
+        if alpha > F::one() {
+            let alpha = a + b;
+            let beta = ((alpha - F::from(2.).unwrap())
+                        / (F::from(2.).unwrap()*a*b - alpha)).sqrt();
+            let gamma = a + F::one() / beta;
+
+            Ok(Beta {
+                a, b, switched_params,
+                algorithm: BetaAlgorithm::BB(BB {
+                    alpha, beta, gamma,
+                })
+            })
+        } else {
+            let alpha = a + b;
+            let beta = F::one() / b;
+            let delta = F::one() + a - b;
+            let kappa1 = delta
+                * (F::from(1. / 18. / 4.).unwrap() + F::from(3. / 18. / 4.).unwrap()*b)
+                / (a*beta - F::from(14. / 18.).unwrap());
```

Comment:
This expression is incorrect and it is due to a typo in the paper. It should be
`/ (b*beta - F::from(14. / 72.).unwrap());`

Note that:
1) I changed `a` for `b` because of the previous comments.
2) I changed `14. / 18.` because of the typo in the paper.

About the typo, $\kappa_1 = \tilde{k}_1 / 4$ and $\tilde{k}_1 = \delta (1 + 3b) / (18 a/b - 14)$, so, at the end,
$$
\kappa_1 = \delta ( \frac{1/72 + b (3/72)}{a \beta - (14/72)} )
$$

Overall, since there is many conversions from float to `F`, I would do only one conversion first:
```rust
let constant = F::from(1./72.).unwrap();
```
and then use it in the expression for `kappa1` in the following way:
```rust
let kappa1 = delta
    * (constant  + F::from(3.).unwrap() * constant * a)
    / (b*beta - F::from(14.).unwrap() * constant);
```
This way, you perform only one division. (feel free to rename the `constant`)

Reply:
I believe the formula is correct? The division by 4 should not affect the denominator.

I think there is only one division at run time, because the other divisions can be performed at compile time.

Reply:
Yes!! My mistake, the 14 in the denominator is okay!! Sorry.

And about the division, I did not know. Thank you!